### ISSUE TITLE
question about the bleu results

### ISSUE THREAD

**Issue by wingsyuan** [OpenNMT/OpenNMT-py#1093]
Hi, I use the opennmt-tf and opennmt-py to train  en-zh with  same training dataset, and translate with the same test dataset too,  but when I evaluate the two result, the score are very different below:

the bleu for opennmt-py:
 perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
BLEU = 20.42, 59.7/30.6/18.6/12.3 (BP=0.803, ratio=0.820, hyp_len=226100, ref_len=275682)

the bleu for opennmt-tf:
 perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.en2zh
BLEU = 27.00, 61.7/34.6/22.5/15.8 (BP=0.915, ratio=0.919, hyp_len=253250, ref_len=275682)

Why? Do I miss something? and How to fine tune opennmt-py to get the similar result? thanks very much!

Process using opennmt-py:
python ../OpenNMT-py-master/preprocess.py -train_src ../data/bpe/untest-src-muti.en.token.bpe -train_tgt ../data/bpe/untest-tgt-muti.zh.token.bpe -valid_src ../data/bpe/untest-vali-muti.en.token.bpe -valid_tgt ../data/bpe/untest-vali-muti.zh.token.bpe -save_data data/data

python train.py -data ../data/data -save_model ../test2/model -layers 6 -rnn_size 512 -word_vec_size 512 -transformer_ff 2048 -heads 8 -encoder_type transformer -decoder_type transformer -position_encoding -train_steps 30000  -max_generator_batches 2 -dropout 0.1 -batch_size 4096 -batch_type tokens -normalization tokens  -accum_count 2  -optim adam -adam_beta2 0.998 -decay_method noam -warmup_steps 8000 -learning_rate 2 -max_grad_norm 0 -param_init 0  -param_init_glorot  -label_smoothing 0.1 -valid_steps 10000 -save_checkpoint_steps 10000  -world_size 4 -gpu_ranks 0 1 2 3 &

python ../opennmt-py/translate.py -model model_step_30000.pt -src untest-test-muti.en.token.bpe -output untest-test-muti.zh.pred.txt -replace_unk -verbose -gpu 4

Process using opennmt-tf:

onmt-main train_and_eval --model_type Transformer --config data.yml --auto_config --session_config gpuconfig --num_gpus 4

onmt-main infer --auto_config --config data.yml --features_file ../data/bpe/untest-test-muti.en.token.bpe >untest-test-muti.en2zh

**Comment by vince62s** [comment]
Did you train only 30k steps with onmt-tf also ?

**Comment by wingsyuan** [comment]
Yes, in data.yml ，I set the parameter below:

train:
  train_steps: 30000

**Comment by vince62s** [comment]
hypothesis length is very different between -tf and -py make sure output files are the same in number of lines, maybe something went wrong during translate.py

**Comment by guillaumekln** [comment]
Yes, the default maximum length during the preprocessing is different:

* OpenNMT-py: 50
* OpenNMT-tf: 100

That's the only big difference I see for now.

**Comment by wingsyuan** [comment]
Hi,@guillaumekln
Thanks for all your help!

but what does the  the default maximum length mean? and where to modify it?

**Comment by guillaumekln** [comment]
Sentences with a length longer than this value are dropped.

* OpenNMT-py: see the preprocessing options `--src_seq_length` and `--tgt_seq_length`
* OpenNMT-tf: see the training parameters `maximum_features_length` and `maximum_labels_length`

**Comment by wingsyuan** [comment]
Hi,
I modify the command to add  options --src_seq_length and --tgt_seq_length like below :
python ../opennmt-py/preprocess.py -train_src ../data/bpe/untest-src-muti.en.token.bpe -train_tgt ../data/bpe/untest-tgt-muti.zh.token.bpe -valid_src ../data/bpe/untest-vali-muti.en.token.bpe -valid_tgt ../data/bpe/untest-vali-muti.zh.token.bpe -save_data data/py/data --src_seq_length 150 --tgt_seq_length 150

and the bleu for opennmt-py is 25.22, compared to the previous bleu for opennmt-py, Increasing the the default maximum length did help : 

perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
BLEU = 25.22, 62.2/34.2/22.0/15.2 (BP=0.869, ratio=0.877, hyp_len=241844, ref_len=275682):
perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
BLEU = 20.42, 59.7/30.6/18.6/12.3 (BP=0.803, ratio=0.820, hyp_len=226100, ref_len=275682)

but still smaller than the previous  bleu score for opennmt-tf( with the default maximum=100):
perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.en2zh
BLEU = 27.00, 61.7/34.6/22.5/15.8 (BP=0.915, ratio=0.919, hyp_len=253250, ref_len=275682)

Is there any other way to fine tune ? thanks very much!

**Comment by wingsyuan** [comment]
ok ,thanks very much for your help!@vince62s

**Comment by vince62s** [comment]
closing this for now, but if you get more results showing comparability between -py and -tf, we are always interested in making things as close as possible.

**Comment by ZacharyWaseda** [comment]
> Hi,
> I modify the command to add options --src_seq_length and --tgt_seq_length =150 like below :
> 
> python ../opennmt-py/preprocess.py -train_src ../data/bpe/untest-src-muti.en.token.bpe -train_tgt ../data/bpe/untest-tgt-muti.zh.token.bpe -valid_src ../data/bpe/untest-vali-muti.en.token.bpe -valid_tgt ../data/bpe/untest-vali-muti.zh.token.bpe -save_data data/py/data --src_seq_length 150 --tgt_seq_length 150
> 
> and the bleu score is 25.22, compare to the previous bleu for opennmt-py, Increasing the the default maximum length did help :
> 
> perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
> BLEU = 25.22, 62.2/34.2/22.0/15.2 (BP=0.869, ratio=0.877, hyp_len=241844, ref_len=275682):
> perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
> BLEU = 20.42, 59.7/30.6/18.6/12.3 (BP=0.803, ratio=0.820, hyp_len=226100, ref_len=275682)
> 
> but still smaller than the previous bleu score for opennmt-tf( with the default maximum=100), and the hyp_len is very different, what does the hyp_len mean?
> perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.en2zh
> BLEU = 27.00, 61.7/34.6/22.5/15.8 (BP=0.915, ratio=0.919, hyp_len=253250, ref_len=275682)
> 
> Is there any other way to fine tune ? thanks very much!

hi, did you find out why opennmt-py still smaller than the previous bleu score for opennmt-tf( with the default maximum=100)? I also meet this problem. thx

**Comment by ZacharyWaseda** [comment]
> Hi,
> I modify the command to add options --src_seq_length and --tgt_seq_length =150 like below :
> 
> python ../opennmt-py/preprocess.py -train_src ../data/bpe/untest-src-muti.en.token.bpe -train_tgt ../data/bpe/untest-tgt-muti.zh.token.bpe -valid_src ../data/bpe/untest-vali-muti.en.token.bpe -valid_tgt ../data/bpe/untest-vali-muti.zh.token.bpe -save_data data/py/data --src_seq_length 150 --tgt_seq_length 150
> 
> and the bleu score is 25.22, compare to the previous bleu for opennmt-py, Increasing the the default maximum length did help :
> 
> perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
> BLEU = 25.22, 62.2/34.2/22.0/15.2 (BP=0.869, ratio=0.877, hyp_len=241844, ref_len=275682):
> perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.zh.pred.txt
> BLEU = 20.42, 59.7/30.6/18.6/12.3 (BP=0.803, ratio=0.820, hyp_len=226100, ref_len=275682)
> 
> but still smaller than the previous bleu score for opennmt-tf( with the default maximum=100), and the hyp_len is very different, what does the hyp_len mean?
> perl ../opennmt-py/tools/multi-bleu.perl untest-test-muti.zh.token.bpe < untest-test-muti.en2zh
> BLEU = 27.00, 61.7/34.6/22.5/15.8 (BP=0.915, ratio=0.919, hyp_len=253250, ref_len=275682)
> 
> Is there any other way to fine tune ? thanks very much!

hi @wingsyuan, finally did you reproduce the same results as openNMT-tf? Now I meet the same problem. thx in advance!
